src/main/java/com/google/sps/data/DataStoreManager.java

```diff
@@ -0,0 +1,342 @@
+  }
+
+  /**
+    * Stores the static data of a single stage in datastore as an entity
+    * @param game an Stage variable representing a single instance of a stage.
+  */
+  public void storeStage(Stage stage) {
+    Entity stageEntity = new Entity("Stage", stage.getStageID());
+    stageEntity.setProperty("stageNumber", stage.getStageNumber());
+    stageEntity.setProperty("key", stage.getKey());
+    stageEntity.setProperty("startingHint", stage.getStartingHint());
+    stageEntity.setProperty("latitude", stage.getStartingLocation().getLatitude());
+    stageEntity.setProperty("longitude", stage.getStartingLocation().getLongitude());
+    for (Hint hint : stage.getHints()) {
+      storeHint(hint);
+    }
+    ArrayList<String> hints;
+    for (Hint hint : stage.getHints()) {
+      hints.add(hint.getHintID());
+    }
+    stageEntity.setProperty("hints", hints);
+    datastore.put(stageEntity);
+  }
+  
+  /**
+    * Updates an entity of a single stage data in datastore
+    * @param stage a Stage variable representing a single instance of a stage.
+  */
+  public void updateStage(Stage stage) {
+    Key stageEntityKey = KeyFactory.createKey("Stage", stage.getStageID());
+    Entity stageEntity;
+    try {
+      stageEntity = datastore.get(stageEntityKey);
+    }
+    catch(Exception e) {
+      return;
+    }
+      
+    stageEntity.setProperty("stageNumber", stage.getStageNumber());
+    stageEntity.setProperty("key", stage.getKey());
+    stageEntity.setProperty("startingHint", stage.getStartingHint());
+    stageEntity.setProperty("latitude", stage.getStartingLocation().getLatitude());
+    stageEntity.setProperty("longitude", stage.getStartingLocation().getLongitude());
+    for (Hint hint : stage.getHints())
+    {
+      updateHint(hint);
+    }
+    ArrayList<String> hints;
+    for (Hint hint : stage.getHints()) {
+      hints.add(hint.getHintID());
+    }
+    stageEntity.setProperty("hints", hints);
+    datastore.put(stageEntity);
+  }
+  
+  /**
+    * Retrieves a single stage entity from the datastore.
+    * @param stageID the unique code used to identify this specific stage.
+    * @return a Stage object with the properties specified in the Builder.
+  */
+  public Game retrieveStage(String stageID) {
+    Key stageEntityKey = KeyFactory.createKey("Stage", stageID);
+    Entity stageEntity;
+    try {
+      stageEntity = datastore.get(stageEntityKey);
+    }
+    catch(Exception e) {
+      return;
+    }
+
+    int stageNumber = (int) stageEntity.getProperty("stageNumber");
+    String key = (String) stageEntity.getProperty("key");
+    String startingHint = (String) stageEntity.getProperty("startingHint");
+    double latitude = (double) stageEntity.getProperty("latitude");
+    double longitude = (double) stageEntity.getProperty("longitude");
+    Coordinates startingLocation;
+    startingLocation = new Coordinates(latitude, longitude);
+    ArrayList<String> hintIDs = stageEntity.getProperty("hints");
```

Comment:
I think you need to cast this to `ArrayList<String>`. So this would be
`ArrayList<String> hintIDs = (ArrayList<String>)stageEntity.getProperty("hints");`

Reply:
done